platforms/core-configuration/configuration-problems-base/src/main/kotlin/org/gradle/internal/configuration/problems/PropertyProblem.kt

```diff
@@ -142,149 +146,120 @@ sealed class PropertyTrace {
+            text(" of type ")
+            reference(type.name)
+        }
     }
 
     data class Bean(
         val type: Class<*>,
         val trace: PropertyTrace
     ) : PropertyTrace() {
-        override val containingUserCode: String
-            get() = trace.containingUserCode
-
+        override val containingUserCodeMessage: StructuredMessage
+            get() = trace.containingUserCodeMessage
         override fun toString(): String = asString()
+        override fun describe(builder: StructuredMessage.Builder) = builder.apply {
+            reference(type.name)
+            text(" bean found in ")
+        }
     }
 
     data class Property(
         val kind: PropertyKind,
         val name: String,
         val trace: PropertyTrace
     ) : PropertyTrace() {
-        override val containingUserCode: String
-            get() = trace.containingUserCode
-
+        override val containingUserCodeMessage: StructuredMessage
+            get() = trace.containingUserCodeMessage
         override fun toString(): String = asString()
+        override fun describe(builder: StructuredMessage.Builder) = builder.apply {
+            text("$kind ")
+            reference(name)
+            text(" of ")
+        }
     }
 
     data class Project(
         val path: String,
         val trace: PropertyTrace
     ) : PropertyTrace() {
-        override val containingUserCode: String
-            get() = trace.containingUserCode
-
+        override val containingUserCodeMessage: StructuredMessage
+            get() = trace.containingUserCodeMessage
         override fun toString(): String = asString()
+        override fun describe(builder: StructuredMessage.Builder) = builder.apply {
+            text("project ")
+            reference(path)
+            text(" in ")
+        }
     }
 
     data class SystemProperty(
         val name: String,
         val trace: PropertyTrace
     ) : PropertyTrace() {
-        override val containingUserCode: String
-            get() = trace.containingUserCode
-
+        override val containingUserCodeMessage: StructuredMessage
+            get() = trace.containingUserCodeMessage
         override fun toString(): String = asString()
+        override fun describe(builder: StructuredMessage.Builder) = builder.apply {
+            text("system property ")
+            reference(name)
+            text(" set at ")
+        }
     }
 
-    abstract override fun toString(): String
-
     abstract override fun equals(other: Any?): Boolean
 
     abstract override fun hashCode(): Int
 
+    abstract override fun toString(): String
+
     /**
      * The shared logic for implementing `toString()` in subclasses.
      */
     protected
     fun asString(): String =
-        StringBuilder().apply {
+        StructuredMessage.Builder().apply {
             sequence.forEach {
-                appendStringOf(it)
+                it.describe(this)
             }
-        }.toString()
+        }.build().asString('`')
```

Comment:
@alllex Here and https://github.com/gradle/gradle/pull/29597/files#diff-8fe7439cf5cc3e70cb39f910cb0c3cefd945ed2a7b76a1011ed7989a21b28a06R257 are callers using backticks (re: your question on slack)

Reply:
I think we should replace the backticks with single quotes in a follow up PR

Reply:
Re: discussion on slack, we have reasons to keep using backticks. Tried to convey some of that as kdoc comments.